src/Sylius/Bundle/GridBundle/Doctrine/PHPCRODM/ExpressionVisitor.php

```diff
@@ -0,0 +1,137 @@
+    private $queryBuilder;
+
+    public function __construct(QueryBuilder $queryBuilder)
+    {
+        $this->queryBuilder = $queryBuilder;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function walkComparison(Comparison $comparison, AbstractNode $parentNode)
+    {
+        $field = $comparison->getField();
+        $value = $comparison->getValue()->getValue(); // shortcut for walkValue()
+
+        switch ($comparison->getOperator()) {
+            case Comparison::EQ:
+                return $parentNode->eq()->field($field)->litreal($value)->end();
+
+            case Comparison::NEQ:
+                return $parentNode->neq()->field($field)->litreal($value)->end();
+
+            case Comparison::LT:
+                return $parentNode->lt()->field($field)->litreal($value)->end();
+
+            case Comparison::LTE:
+                return $parentNode->lte()->field($field)->litreal($value)->end();
+
+            case Comparison::GT:
+                return $parentNode->gt()->field($field)->litreal($value)->end();
+
+            case Comparison::GTE:
+                return $parentNode->gte()->field($field)->litreal($value)->end();
+
+            case Comparison::IN:
+                return $parentNode->in()->field($field)->litreal($value)->end();
+
+            case Comparison::NIN:
+                return $parentNode->notIn()->field($field)->litreal($value)->end();
+
+            case Comparison::CONTAINS:
+                return $parentNode->like()->field($field)->litreal($value)->end();
+
+            default:
+                throw new \RuntimeException("Unknown comparison operator: " . $comparison->getOperator());
+        }
+
+        return $parentNode;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function walkValue(Value $value, OperandFactory $parentNode)
+    {
+        throw new \BadMethodCallException('What is this even used for?');
+        return $parentNode->literal($value->getValue());
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function walkCompositeExpression(CompositeExpression $expr, AbstractNode $abstractNode)
+    {
+        switch($expr->getType()) {
+            case CompositeExpression::TYPE_AND:
+                $baseNode = $parentNode->andX();
+            case CompositeExpression::TYPE_OR:
+                $baseNode = $parentNode->orX();
+            default:
+                throw new \RuntimeException('Unknown composite: ' . $expr->getType());
+        }
+
+        $expressions = $expr->getExpressionList();
+        $node = $baseNode;
+
+        if ($expressionCount = count($expressions) <= 2) {
+            throw new \RuntimeException('Composite expressions must be composed of two or more expressions.');
+        }
+
+        foreach (array_values($expressions) as $index => $expression) {
+            // if it is the last element dispatch the node itself.
+            if ($index === $expressionCount - 1) {
+                $this->dispatch($expression, $node);
```

Comment:
No we should `break` I think - note that this is just very rough code that hasn't been tested.